### PR TITLE
[resty.concurrent] collection of modules for concurrent programming

### PR DESCRIPTION
**Pull request by mikz**
Takes design from concurrent-ruby

### CODE REVIEW

File: spec/resty/concurrent/event_spec.lua [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    after_each(function()
+        for i,thread in ipairs(threads) do
+            ngx.thread.kill(thread)
+            threads[i] = nil
+        end
+    end)
+
+    describe('can be waited by multiple threads', function()
+        it('notifies all threads', function()
```
Comment (davidor) [inline spec/resty/concurrent/event_spec.lua]: I think this title is not descriptive. What about `notifies the main thread` ?

Comment (mikz) [reply spec/resty/concurrent/event_spec.lua]: Yep, copy pasta description.

File: spec/resty/concurrent/executor_spec.lua [hunk 2]
```diff
@@ -0,0 +1,15 @@
+local Executor = require('resty.concurrent.executor')
+
+describe('ImmediateExecutor', function()
+    describe('.from_options', function()
+        it('has default executor', function()
+            assert(Executor.from_options{ })
```
Comment (davidor) [inline spec/resty/concurrent/executor_spec.lua]: Shouldn't we check that the default is the immediate one?

File: spec/resty/concurrent/executor_spec.lua [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+        end)
+
+        it('has passes through executor', function()
+            local executor = { 'something' }
+
+            assert.equal(executor, Executor.from_options{ executor = executor })
+        end)
+    end)
```
Comment (davidor) [inline spec/resty/concurrent/executor_spec.lua]: I think we should also try with strings.  `Executor.from_options { executor = _one_that_exists_ })` and with `_one_that_does_not_exist_`

File: spec/resty/concurrent/executor_spec.lua (hunk 2 above)
Comment (mikz) [reply spec/resty/concurrent/executor_spec.lua]: Yep, why not.

File: gateway/src/apicast/proxy.lua [hunk 4]
```diff
@@ -306,55 +310,28 @@ local function response_codes_data()
   return params
 end
 
-local function handle_timer_post_action(_, self, semaphore, cached_key, backend, ...)
-  local res = util.timer('backend post_action', backend.authrep, backend, ...)
-
-  if not using_sync_post_action then
-    semaphore:post(1)
-  end
+local http_backend = {
```
Comment (davidor) [inline gateway/src/apicast/proxy.lua]: Why is this needed?

Comment (mikz) [reply gateway/src/apicast/proxy.lua]: We can't use the `ngx` backend in timers because it uses subrequests. It can use the `ngx` backend only in `access`/ `rewrite`/ `content` phases.

File: spec/resty/concurrent/safe_task_executor_spec.lua [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+        assert.same({true, { 'some value', 'some other value' }}, { task:execute() })
+    end)
+
+    it('returns failure', function ()
+        local task = SafeTaskExecutor.new(function () return error('some error') end)
+
+        assert.same({false, nil, 'some error'}, { task:execute() })
+    end)
+
```
[the remark this answers is not in the document]
Comment (mikz) [reply spec/resty/concurrent/safe_task_executor_spec.lua]: Yep 👍

File: gateway/src/apicast/proxy.lua (hunk 4 above)
Comment (davidor) [reply gateway/src/apicast/proxy.lua]: :+1:

**Comment by davidor** [comment]
I wrote a few comments about minor things.

Looks very good :+1:

**Comment by mikz** [comment]
@davidor fixed all your comments, thanks! 🥇
